### ISSUE TITLE
`typing_extensions.TYPE_CHECKING` branch ignored and `TCH001` & `TCH002` are raised

### ISSUE THREAD

**Issue by chaburkland** [astral-sh/ruff#8427]
```python
# example.py prior to ruff fix
from __future__ import annotations

import typing_extensions as tp

if tp.TYPE_CHECKING:
    from application.types import something
    from sqlalchemy.orm.session import Session


def foo(a: something) -> Session:
    pass
```
Command: `ruff example.py --fix` (Unable to reproduce with the `--isolated` flag)

The fix applies a strange behavior. It appears that `ruff` is unable to identify the `typing_extensions.TYPE_CHECKING` branch.
```python
# example.py after ruff fix
from __future__ import annotations

import typing_extensions as tp
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from application.types import something
    from sqlalchemy.orm.session import Session

if tp.TYPE_CHECKING:
    pass


def foo(a: something) -> Session:
    pass
```

Version: `ruff 0.0.292`

`pyproject.toml`
```toml
[tool.ruff]
target-version = "py38"
exclude = [
	".git",
	"__pycache__",
	".mypy_cache",
	".ruff_cache",
]
ignore = [
	"B011",     
	"B015",     
	"B023",     
	"B028",     
	"F811",     
	"F841",     
	"PLR0911",  
	"PLR0912",  
	"PLR0913",  
	"PLR0915",  
	"PLW2901",  
	"PLW3301",  
	"RUF001",   
	"RUF002",   
	"RUF003",   
	"RUF005",   
	"RUF012",   
	"UP025",    
	"PLR1711",  
	"PLR2004",  
	"PLR5501",  
	"PIE794",
	"F821",     
	"PLC0414",  
]
select = [
	"F",   
	"W",   
	"UP",  
	"B",   
	"T10", 
	"EXE", 
	"ISC", 
	"ICN", 
	"PIE", 
	"RSE", 
	"TCH", 
	"PGH", 
	"PL",  
	"NPY", 
	"RUF", 
]
```

**Comment by zanieb** [comment]
Thanks for the report!

I've started a fix at https://github.com/astral-sh/ruff/pull/8429 with some notes on what else needs to be done.
